packages/gatsby/src/utils/page-ssr-module/get-server-data-utils.ts

```diff
@@ -0,0 +1,85 @@
+import * as fs from "fs-extra"
+import * as t from "@babel/types"
+import traverse from "@babel/traverse"
+import { babelParseToAst } from "../babel-parse-to-ast"
+
+type ComponentPath = string
+type FunctionText = string
+const prevSource: Map<ComponentPath, FunctionText> = new Map()
+
+/**
+ * Returns true when getServerData source is different from the previous time this function was called
+ */
+export async function getServerDataChanged(
+  filename: ComponentPath
+): Promise<boolean> {
+  // TODO:
+  // if (!isSSRPage(filename)) return false
+
+  const prev = prevSource.get(filename)
+  prevSource.delete(filename)
+
+  if (
+    !(await fs.pathExists(filename)) ||
+    !(await fs.lstat(filename)).isFile()
+  ) {
+    return !!prev
+  }
+  const text = await fs.readFile(filename, `utf8`)
+
+  // Avoid expensive AST checks for pages without getServerData
+  if (!text.includes(`getServerData`)) {
+    return !!prev // if it was set before - means it was changed
+  }
+  const source = extractSource(text, filename)
+  prevSource.set(filename, source)
```

Comment:
Honestly, I would skip this check and if the module is part of an ssr page we should always refresh even if it's a no-op. It will make our lives easier.

For example this example isn't working
```
function getCustomData() {
  return {
    message: "hello"
  }
}

export function getServerData() {
  return {
    props: getCustomData()
  }
}

```

Reply:
> I'll remove this check but it's just trading one dx problem for another (say editing only react component or graphql query may be much slower because of this).

Sure but babel tracking is also expensive so we should go with the simplest solution for now until people complain.

Reply:
Simplified the logic in https://github.com/gatsbyjs/gatsby/pull/33398/commits/09490eca99fd293c71395187bdc090549f066091

Now it rebuilds when a changed file contains `getServerData`. So if you delete it from the file - it won't rebuild the SSR bundle. That's probably fine in develop.